storage/src/adb/any/sync/client.rs

```diff
@@ -0,0 +1,726 @@
+                    next_op_loc = next_op_loc,
+                    remaining_ops = remaining_ops,
+                    batch_size = batch_size.get(),
+                    "Fetching proof and operations"
+                );
+
+                // Get proof and operations from resolver with timing
+                let fetch_start = config.context.current();
+                let (proof, new_operations) =
+                    config.resolver.get_proof(next_op_loc, batch_size).await?;
+                let fetch_end = config.context.current();
+                let fetch_secs = fetch_end
+                    .duration_since(fetch_start)
+                    .unwrap_or_default()
+                    .as_secs_f64();
+                metrics.fetch_duration.observe(fetch_secs);
+
+                let new_operations_len = new_operations.len() as u64;
+
+                // Validate that we didn't get more operations than requested
+                // or that we didn't get an empty proof. We should never get an empty proof
+                // because we will never request an empty proof (i.e. a proof over an empty database).
+                if new_operations_len > batch_size.get() || new_operations_len == 0 {
+                    metrics.invalid_batches_received.inc();
+                    if metrics.invalid_batches_received.get() > config.max_retries {
+                        return Err(Error::MaxRetriesExceeded);
+                    }
+                    return Ok(Client::FetchData {
+                        config,
+                        log,
+                        pinned_nodes,
+                        metrics,
+                    });
+                }
+
+                debug!(
+                    num_ops = new_operations_len,
+                    "Received operations from resolver"
+                );
+
+                // Verify the proof is valid over the given operations with timing
+                let verification_start = config.context.current();
+                let proof_valid = adb::any::Any::<E, K, V, H, T>::verify_proof(
+                    &mut config.hasher,
+                    &proof,
+                    next_op_loc,
+                    &new_operations,
+                    &config.target_hash,
+                );
+                let verification_end = config.context.current();
+                let verify_secs = verification_end
+                    .duration_since(verification_start)
+                    .unwrap_or_default()
+                    .as_secs_f64();
+                metrics.proof_verification_duration.observe(verify_secs);
+
+                if !proof_valid {
+                    debug!("Proof verification failed, retrying");
+                    metrics.invalid_batches_received.inc();
+                    if metrics.invalid_batches_received.get() > config.max_retries {
+                        return Err(Error::MaxRetriesExceeded);
+                    }
+                    config.resolver.notify_failure();
```

Comment:
I think we should instead have the resolver return a `oneshot` with the response from get_proof that you can call `false` on?

Reply:
The current approach will make it hard to map which response failed when we add concurrency.